edb/edgeql/compiler/normalization.py

```diff
@@ -0,0 +1,539 @@
+#
+# Copyright 2020-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either nodeess or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+"""EdgeQL expression normalization functions."""
+
+
+from __future__ import annotations
+from typing import *
+
+import copy
+import functools
+
+from edb.common.ast import base
+from edb.edgeql import ast as qlast
+from edb.schema import schema as s_schema
+from edb.schema import name as sn
+
+
+@functools.singledispatch
+def normalize(
+    node: Any,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> Any:
+    raise ValueError(f'normalize: cannot handle {node!r}')
+
+
+def _normalize_generic_field(
+    node: qlast.Base,
+    field: str,
+    value: Any,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> None:
+    if isinstance(value, qlast.Base):
+        setattr(
+            node,
+            field,
+            normalize(
+                value,
+                schema=schema,
+                modaliases=modaliases,
+                ignorenames=ignorenames,
+            )
+        )
+    elif isinstance(value, list):
+        if value and isinstance(value[0], qlast.Base):
+            setattr(
+                node,
+                field,
+                [
+                    normalize(
+                        el,
+                        schema=schema,
+                        modaliases=modaliases,
+                        ignorenames=ignorenames,
+                    ) for el in value
+                ]
+            )
+
+
+@normalize.register
+def normalize_Base(
+    node: qlast.Base,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> qlast.Base:
+    node = copy.copy(node)
+
+    for field, value in base.iter_fields(node):
+        _normalize_generic_field(
+            node,
+            field,
+            value,
+            schema=schema,
+            modaliases=modaliases,
+            ignorenames=ignorenames,
+        )
+
+    return node
+
+
+@normalize.register
+def normalize_DDL(
+    node: qlast.DDL,
+    *,
+    schema: s_schema.Schema,
+    modaliases: Mapping[Optional[str], str],
+    ignorenames: Set[str] = frozenset(),
+) -> qlast.DDL:
+    raise ValueError(f'normalize: cannot handle {node!r}')
```

Comment:
Let's use `AssertionError` for unexpected code paths.

Reply:
Sure, I've changed `ValueError` to `AssertionError`.